File: python/registream-core/tests/test_citation.py
"""Unit tests for registream.citation."""

from __future__ import annotations

from importlib.metadata import version

from registream.citation import cite, cite_bibtex


# ─── cite() ────────────────────────────────────────────────────────────────────


def test_cite_default_includes_authors_and_url() -> None:
    text = cite()
    assert "Clark, J. & Wen, J." in text
    assert "https://registream.org" in text
    assert "RegiStream" in text


def test_cite_default_is_versioned() -> None:
    text = cite()
    installed = version("registream-core")
    assert f"Version {installed}" in text


def test_cite_unversioned_omits_version_string() -> None:
    text = cite(versioned=False)
    assert "Version" not in text
    assert "Clark, J. & Wen, J." in text
    assert "https://registream.org" in text


def test_cite_returns_string_not_none() -> None:
    assert isinstance(cite(), str)
    assert cite() != ""


# ─── cite_bibtex() ─────────────────────────────────────────────────────────────


def test_cite_bibtex_has_required_fields() -> None:
    bib = cite_bibtex()
    assert "@software{clark2024registream," in bib
    assert "author" in bib
    assert "title" in bib
    assert "version" in bib
    assert "year" in bib
    assert "url" in bib


def test_cite_bibtex_includes_installed_version() -> None:
    installed = version("registream-core")
    assert f"version = {{{installed}}}" in cite_bibtex()


def test_cite_bibtex_brace_balance() -> None:
    """The BibTeX must have balanced braces; common f-string mistake."""
    bib = cite_bibtex()
    assert bib.count("{") == bib.count("}")

File: r/tests/testthat/test-citation.R
# rs_cite() / rs_cite_bibtex() pull from the generated _citation_data.R.

test_that("rs_cite() returns APA text with author + title", {
  out <- rs_cite(versioned = FALSE)
  expect_match(out, "Clark")
  expect_match(out, "Wen")
  expect_match(out, "RegiStream:")
  expect_match(out, "https://registream.org", fixed = TRUE)
})


test_that("rs_cite() includes installed registream version by default", {
  out <- rs_cite(versioned = TRUE)
  v <- as.character(utils::packageVersion("registream"))
  expect_match(out, sprintf("Version %s", v), fixed = TRUE)
})


test_that("rs_cite_bibtex() returns a @software entry", {
  out <- rs_cite_bibtex(versioned = TRUE)
  expect_match(out, "^@software")
  expect_match(out, "clark2024registream")
})
